packages/logger/src/logger.ts

```diff
@@ -137,23 +166,102 @@ export class Logger {
     this._logHandler = val;
   }
 
+  /**
+   * The optional, additional, user-defined log handler for the Logger instance.
+   */
+  private _userLogHandler: LogHandler | null = null;
+  get userLogHandler(): LogHandler | null {
+    return this._userLogHandler;
+  }
+  set userLogHandler(val: LogHandler | null) {
+    this._userLogHandler = val;
+  }
+
   /**
    * The functions below are all based on the `console` interface
    */
 
   debug(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.DEBUG, ...args);
     this._logHandler(this, LogLevel.DEBUG, ...args);
   }
   log(...args: unknown[]): void {
+    this._userLogHandler &&
+      this._userLogHandler(this, LogLevel.VERBOSE, ...args);
     this._logHandler(this, LogLevel.VERBOSE, ...args);
   }
   info(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.INFO, ...args);
     this._logHandler(this, LogLevel.INFO, ...args);
   }
   warn(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.WARN, ...args);
     this._logHandler(this, LogLevel.WARN, ...args);
   }
   error(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.ERROR, ...args);
     this._logHandler(this, LogLevel.ERROR, ...args);
   }
 }
+
+export function setLogLevel(level: LogLevelString | LogLevel): void {
+  const newLevel = typeof level === 'string' ? levelStringToEnum[level] : level;
+  instances.forEach(inst => {
+    inst.logLevel = newLevel;
+  });
+}
+
+export function setUserLogHandler(
+  logCallback: LogCallback | null,
+  options?: LogOptions
+): void {
+  if (logCallback !== null && typeof logCallback !== 'function') {
+    throw new TypeError(
+      'First argument to `onLog` must be null or a function.'
```

Comment:
This function should not know about `onLog` that's defined on firebase namespace. Instead, we should catch this error in `@firebase/app` and wrap it in a `FirebaseError` with the correct message.

Reply:
Moved.